### PR TITLE
WIP: Update readme 17

### PR DESCRIPTION
**Pull request by groovecoder**
Based on https://github.com/mozilla/blurts-server/pull/13 so that should merge first. Just putting this up to get first comments.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,81 +1,78 @@
...
 
-```
-{
-  "username": "<username>",
-  "password": "<password>"
-}
+1. Source the `.env` file to set required environment variables:
 
-```
+        source .env
```
Comment (pdehaan) [inline README.md]: Do we need to do this? I think the WIP server.js does one of these: `require("dotenv").load();`, which I _thought_ handled importing .env files and merging w/ env vars automatically.

File: README.md [hunk 2]
```diff
@@ -1,81 +1,78 @@
...
 
+You can set and source these via the `.env` file, or set them directly:
 ```
-$ cd BreachAlerts/client-addon
-$ ./package.sh # This creates a client-addon.xpi file in the dist/ folder
+export DEBUG_DUMMY_SMTP=0
+export SMTP_HOST=<your-smtp-host>
+export SMTP_PORT=<your-smtp-port>
+export SMTP_USERNAME=<your-username>
+export SMTP_PASSWORD=<your-password
```
Comment (pdehaan) [inline README.md]: nit: missing trailing ">"

File: server.js [hunk 3]
```diff
@@ -223,10 +224,9 @@ if (process.env.DEBUG_DUMMY_SMTP) {
...
 }
 
-if (gTransporter) {
-  app.listen(port, function() {
-    console.log("Listening on " + port);
-  });
-} else {
-  console.error("SMTP credentials couldn't be read from the environment, exiting.");
+if (!gTransporter) {
+  console.log("SMTP config unavailble. Email features will not work.")
```
Comment (pdehaan) [inline server.js]: typo: "unavailable"

Comment (pdehaan) [reply server.js]: Also, what happens when we set `gTransporter` to null above on L221?
I think it'll blow up above on L123 where we try and do `gTransporter.sendMail(...)` and we'll need to maybe set the `gTransporter` object to a dummy object like on L197-201.

File: server.js [hunk 4]
```diff
@@ -200,14 +200,15 @@ if (process.env.DEBUG_DUMMY_SMTP) {
     },
   };
 } else {
-  console.log("Attempting to get SMTP credentials from environment...");
+  console.log("Attempting to get SMTP config from environment...");
   kSMTPUsername = process.env.SMTP_USERNAME;
   let password = process.env.SMTP_PASSWORD;
+  let SMTP_host = process.env.SMTP_HOST;
+  let SMTP_port = process.env.SMTP_PORT;
```
Comment (pdehaan) [inline server.js]: Possibly unrelated, but these are starting to feel confusingly inconsistent. Variables are either `password`, `SMTP_port`, or `kSMTPUsername`. Not sure if they should be `smtpUsername`, `smtpPassword`, `smtpHost`, and `smtpPort` for consistency (and changed globally, in the case of `kSMTPUsername`).

File: server.js [hunk 5]
```diff
@@ -57,8 +131,67 @@ app.post("/user/breached", function(req, res) {
...
+// TODO: remove this
+app.get("/test", function(req, res) {
+  res.send(req.body);
+})
+
 var port = process.env.PORT || 6060;
 
+// Allow a debug mode that will send JSON back to the client instead of sending emails.
 if (process.env.DEBUG_DUMMY_SMTP) {
   console.log("Running in dummp SMTP mode, /user/breached will send a JSON response instead of sending emails.");
```
Comment (pdehaan) [inline server.js]: typo: "dummy"

File: README.md (hunk 1 above)
Comment (nhnt11) [reply README.md]: This is correct.

File: .env-dist [hunk 6]
```diff
@@ -6,3 +6,7 @@ OAUTH_PROFILE_URI="https://stable.dev.lcip.org/profile/v1/profile"
 OAUTH_TOKEN_URI="https://oauth-stable.dev.lcip.org/v1/token"
 SERVER_URL=http://localhost:6060
 COOKIE_SECRET=3895d33b5f9730f5eb2a2067fe0a690e298f55f5e382c032fd3656863412
+SMTP_HOST=""
+SMTP_PORT=
+SMTP_USERNAME=""
+SMTP_PASSWORD=""
```
[the remark this answers is not in the document]
Comment (nhnt11) [reply .env-dist]: This seemed like a good point at first, but I think we need to allow a separate PORT to allow this to be easily deployed through a reverse proxy or some such thing. Server URL is still important for things like constructing a redirect URL for OAuth.

File: server.js (hunk 3 above)
Comment (nhnt11) [reply server.js]: I'd prefer simply failing and exiting in this case for now.

File: server.js (hunk 4 above)
Comment (nhnt11) [reply server.js]: Additionally, we need to null check these in the next line.

File: README.md [hunk 7]
```diff
@@ -1,81 +1,92 @@
...
-}
+1. Source the `.env` file to set required environment variables:
 
-```
+        source .env
+
+2. Run the server:
 
-NOTE: Currently the Gmail SMTP server is hardcoded, so you'll need to modify that in server.js if you're using something else.
+        node server.js
```
Comment (nhnt11) [inline README.md]: npm start is the preferred way to run the server, I'd say. The "start" script is defined (to run node server.js) in package.json, and if we want to have a more sophisticated run script we can just modify that without changing the instructions.

File: README.md [hunk 8]
```diff
@@ -1,81 +1,92 @@
...
 
-```
-{
-  "username": "<username>",
-  "password": "<password>"
-}
+1. Source the `.env` file to set required environment variables:
 
-```
+        source .env
```
Comment (nhnt11) [inline README.md]: We don't need this line, dotenv takes care of it in the beginning of server.js

File: README.md [hunk 9]
```diff
@@ -1,81 +1,92 @@
...
+* [node](https://nodejs.org/) (with npm)
 
-## Privacy Concerns
+### Install
 
-The third goal brings up some privacy concerns, since users would need to supply an email address to receive notifications. Who is the custodian of this data? Can we avoid sending user data to haveibeenpwned.com? Can we still offer useful functionality to users who opt out of subscribing their email address? While the project is still in infancy, the idea is to offer as much utility as possible while respecting the user's privacy.
+1. clone and change to the directory:
 
-## Try it (basic testing instructions for Unix OSes as of Nov 23 2017)
+        git clone https://github.com/nhnt11/blurts-server.git
```
Comment (nhnt11) [inline README.md]: As pdehaan says, the repository is in the mozilla org now.

File: README.md (hunk 2 above)
Comment (nhnt11) [reply README.md]: echoing pdehaan:
nit: missing trailing ">"

File: README.md [hunk 10]
```diff
@@ -1,81 +1,92 @@
...
 
-Prerequisites: currently Firefox 58.0 or 59.0.
+To use Firefox Accounts, you'll need to [create an FxA Oauth Client](https://oauth-stable.dev.lcip.org/console/clients)
+and then set some `OAUTH` config values.
 
+You can set and source these via the `.env` file, or set them directly:
 ```
-$ cd BreachAlerts/client-addon
-$ ./package.sh # This creates a client-addon.xpi file in the dist/ folder
+OAUTH_CLIENT_ID=<your-fxa-oauth-client-id>
```
Comment (nhnt11) [inline README.md]: We can probably mention the existing dev OAuth relier account. I found precedent for exposing staging/development credentials elsewhere (can't remember now, will try to find it), so I put our client_id and client_secret in .env_dist.

File: server.js (hunk 4 above)
Comment (nhnt11) [reply server.js]: Let's rename these variables to just host and port. They're single-use, we can keep it simple. Also don't forget to null check them in the line below.

File: server.js [hunk 11]
```diff
@@ -223,10 +224,9 @@ if (process.env.DEBUG_DUMMY_SMTP) {
...
-if (gTransporter) {
-  app.listen(port, function() {
-    console.log("Listening on " + port);
-  });
-} else {
-  console.error("SMTP credentials couldn't be read from the environment, exiting.");
+if (!gTransporter) {
+  console.log("SMTP config unavailble. Email features will not work.")
 }
+app.listen(port, function() {
```
Comment (nhnt11) [inline server.js]: Let's just not run the server if we failed to get valid SMTP config, for now. In the future we can account for invalid config in the APIs handlers that use SMTP.

**Comment by pdehaan** [comment]
@groovecoder After you rebase/address comments and we're safe to merge this, I'll take another stab at my glorious ESLint PR. 🔪

File: server.js (hunk 3 above)
Comment (groovecoder) [reply server.js]: I ran into this when I tried to deploy the server to a new heroku app to test Firefox Accounts. Requiring a fully-configured SMTP backing service means it's impossible to work on any other part of the app - e.g., Firefox Accounts, front-end web pages, REST API, scheduled jobs/processes - unless you first configure an SMTP backing service.

File: README.md (hunk 10 above)
Comment (groovecoder) [reply README.md]: Is that client configured to redirect back to localhost:6060 properly? When I used the values in `.env-dist`, I got "invalid request parameter" at http://localhost:6060/oauth/redirect?state=2e2...6d4&code=972...8f5&action=signin when I signed in with FxA.

**Comment by groovecoder** [comment]
This is rebased and almost all the comments are addressed.

File: README.md (hunk 10 above)
Comment (nhnt11) [reply README.md]: It should "just work". The FxA local dev console specifically says that it allows localhost redirects for ease of development. Thanks though, I'll look into this.

File: server.js (hunk 3 above)
Comment (nhnt11) [reply server.js]: But that's the purpose of the dummy SMTP mode flag. I added that for a "clean" way to explicitly opt-in to a "dev mode" of sorts that doesn't require an SMTP service. Let me know what you think, or if I'm missing something.

**Pull request merged**